src/VisualStudio/Core/Impl/CodeModel/ProjectCodeModelFactory.cs

```diff
@@ -66,21 +65,35 @@ internal sealed class ProjectCodeModelFactory : IProjectCodeModelFactory
             _visualStudioWorkspace.WorkspaceChanged += OnWorkspaceChanged;
         }
 
-        private System.Threading.Tasks.Task ProcessNextDocumentBatchAsync(
+        private async System.Threading.Tasks.Task ProcessNextDocumentBatchAsync(
             ImmutableArray<DocumentId> documentIds, CancellationToken cancellationToken)
         {
+            // This logic preserves teh previous behavior we had with IForegroundNotificationService.
+            // Specifically, we don't run on the UI thread for more than 15ms at a time.  And once we 
+            // have, we wait 50ms before continuing.  These constants are just what we defined from
+            // legacy, and otherwise have no special meaning.
+            const int MaxTimeSlice = 15;
+            const int DelayBetweenProcessing = 50;
+
+            await _threadingContext.JoinableTaskFactory.SwitchToMainThreadAsync(cancellationToken);
+
+            var stopwatch = SharedStopwatch.StartNew();
             foreach (var documentId in documentIds)
             {
-                // Now, enqueue foreground work to actually process these documents in a serialized and incremental
-                // fashion.  FireEventsForDocument will actually limit how much time it spends firing events so that it
-                // doesn't saturate  the UI thread.
-                _notificationService.RegisterNotification(
-                    () => FireEventsForDocument(documentId),
-                    _listener.BeginAsyncOperation("CodeModelEvent"),
-                    cancellationToken);
+                do
+                {
+                    // Keep firing events for this doc, as long as we haven't exceeded the max amount
+                    // of waiting time, and there's no user input that should take precedence.
+                    if (stopwatch.Elapsed.Ticks > MaxTimeSlice || IsInputPending())
+                    {
+                        await Task.Delay(DelayBetweenProcessing, cancellationToken).ConfigureAwait(true);
+                        stopwatch = SharedStopwatch.StartNew();
+                    }
+                }
+                while (FireEventsForDocument(documentId));
```

Comment:
@sharwell this should actually give the same semantics as before.  However, if we're worried about this, we could keep IFGService, but just move it into the CodeModel project.